File: elixir/ch1/ex113.ex
# cd("c:/Users/Yohan/IdeaProjects/sicp/elixir/ch1")
# Load with import_file("ex113.ex")
# Run with Ex113.fib(n)
defmodule Ex113 do
  @moduledoc false
  def psi, do: (1 + :math.sqrt(5)) / 2
  def phi, do: (1 - :math.sqrt(5)) / 2

  def fib(n) when n < 2, do: n
  def fib(n), do: round((:math.pow(psi(), n) - :math.pow(phi(), n)) / :math.sqrt(5))
end

File: haskell/ch1/Ex113.hs
-- :cd C:\Users\Yohan\IdeaProjects\sicp\haskell\ch1
-- :load "Ex112.hs"
-- Run with fib 15 (should equal 610)

module Ex113 where

-- return nth Fibonacci number
psi :: Int -> Float
psi n = ((1.0 + sqrt 5) / 2) ** fromIntegral n
phi :: Int -> Float
phi n = ((1.0 - sqrt 5) / 2) ** fromIntegral n

fib :: Int -> Int
fib n | n < 2 = n
fib n = round (((psi n) - (phi n)) / sqrt(5))
